### DISCUSSION TITLE
issue: Errors are not removed from array fields

### DISCUSSION THREAD

**Discussion by marcospassos** [react-hook-form/react-hook-form#9875] [Q&A]
### Version Number

7.43.0/8-next.4

### Codesandbox/Expo snack

https://codesandbox.io/s/react-hook-form-array-fields-with-validation-yup-forked-w9kye1?file=/src/index.js

### Steps to reproduce

1. Click on "Add friend" 3 times and see an error is shown
2. Click on "Remove friend" and see the error is still shown

### Expected behaviour

Revalidate the form once the item is removed to maintain consistency with the `append` behavior.

Notice that using `trigger` is not an option since it will trigger the validation of all nested fields, which is undesired. In our case, we've deeply nested forms, and doing that results in a terrible UX.

https://user-images.githubusercontent.com/943036/216787017-1fa5f3a5-a766-4c91-8e73-5d73076e5a76.mov

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
Please use the following code to debug the issue first:

```tsx
resolver: async (data, context, options) => {
  // you can debug your validation schema here
  console.log('formData', data)
  console.log('validation result', await anyResolver(schema)(data, context, options))
  return anyResolver(schema)(data, context, options)
},
```

https://react-hook-form.com/api/useform

![Screenshot 2023-02-05 at 9 46 01 am](https://user-images.githubusercontent.com/10513364/216792644-ae1c1876-948c-4b15-87a5-2fd73f83aaa8.png)

**Comment by marcospassos** [comment]
I've updated the example with this debugging part. As you can see below, the schema correctly reports no error when removing an item, but RHF still shows the error:

![image](https://user-images.githubusercontent.com/943036/216792873-b385bf32-65b4-42bf-a51e-77b834461af6.png)

**Comment by bluebill1049** [comment]
https://codesandbox.io/s/react-hook-form-array-fields-with-validation-yup-forked-ztx800?file=/src/index.js

Please take a look at the CSB, I have switched to the latest version and let me know if I am missing something


https://user-images.githubusercontent.com/10513364/216793395-cb6c793d-2b8e-47f8-ab96-56c10769e3b2.mov

**Comment by bluebill1049** [comment]
related issue: https://github.com/react-hook-form/react-hook-form/issues/9868

**Comment by marcospassos** [comment]
The problem is still present in your CSB fork:

https://user-images.githubusercontent.com/943036/216824020-04c00127-d06f-4759-a606-2ee5edf2bb68.mov

Notice that the error correctly appears after adding the third element (the max is 2), but after removing it, the error remains. It looks like the `remove` isn't triggering the revalidation on the array field.

**Comment by bluebill1049** [comment]
<img width="579" alt="Screenshot 2023-02-06 at 7 12 29 am" src="https://user-images.githubusercontent.com/10513364/216842447-974b7781-682a-4e3a-b607-c0e2229f67e1.png">

Your schema is still returning errors even with 2 items.

**Comment by marcospassos** [comment]
It's returning the error for the right fields: the array elements. There's no error for the array itself, but the RHF still shows an error at the array level:

![image](https://user-images.githubusercontent.com/943036/216847005-6a963d0b-f0b2-49cd-8d2c-453a61296bf3.png)

**Comment by bluebill1049** [comment]
Refer to the same reply: https://github.com/react-hook-form/react-hook-form/issues/9868#issuecomment-1418270268

**Comment by marcospassos** [comment]
Sorry, I still don't get it.

Are you saying that RHF showing an error that no longer exists is right? As I have shown, the last revalidation resulted in no error at the array level, but it still keeps a state with an error from the past. 

Even using the `dirtyFields` or `touchedFields`, I can't workaround this issue since the fields are touched and dirty, but the error doesn't exist - the state that's outdated.

**Comment by bluebill1049** [comment]
> Are you saying that RHF showing an error that no longer exists is right? As I have shown, the last revalidation resulted in no error at the array level, but it still keeps a state with an error from the past.

No, The error is still there under the field array key `firends`.

**Comment by marcospassos** [comment]
No, there is no error under the field array key, please see:

https://user-images.githubusercontent.com/943036/216847735-2bdd4d0b-85f9-4128-adc2-aa220d71a69a.mov

**Comment by bluebill1049** [comment]
https://codesandbox.io/s/react-hook-form-array-fields-with-validation-yup-forked-ztx800?file=/src/index.js

https://user-images.githubusercontent.com/10513364/216848063-40929aeb-d1ac-4752-b9bb-8013eb7f4ed1.mov

- 3-row trigger error
- remove 2 still have errors from the schema
- error is hence not removed

**Comment by marcospassos** [comment]
So, are you saying that an error from a parent field will only be removed once all nested field errors have been fixed, even though it no longer has errors?

**Comment by bluebill1049** [comment]
https://github.com/react-hook-form/react-hook-form/issues/9868#issuecomment-1418270268

> pick up the errors by field array key

**Comment by marcospassos** [comment]
Sorry, I didn't get what you meant by "pick up the errors by field array key".

Do I have access to the path of the field that triggered the revalidation from the resolver?

**Comment by bluebill1049** [comment]
> Do I have access to the path of the field that triggered the revalidation from the resolver?

I think you could try that which is what i commented in the above rely as well.

> revalidate the schema (you may be able to do a custom resolver here, to resolve your error, so meet your UX logic)

**Comment by bluebill1049** [comment]
I think we do return the validation field path. Sorry on my phone now reply would be limited.

**Comment by marcospassos** [comment]
No matter what I do, the error doesn't disappear. Even not returning any other error, RHF still shows the error:
https://codesandbox.io/s/react-hook-form-array-fields-with-validation-yup-forked-w9kye1?file=/src/index.js

https://user-images.githubusercontent.com/943036/216850373-cc922676-ee94-4a0b-b256-2f9d89bd9a33.mov

**Comment by bluebill1049** [comment]
https://codesandbox.io/s/react-hook-form-array-fields-with-validation-yup-forked-ztx800?file=/src/index.js

```tsx
import React from "react";
import ReactDOM from "react-dom/client";
import { useFieldArray, useForm } from "react-hook-form";
import * as yup from "yup";
import { yupResolver } from "@hookform/resolvers/yup";
import "./styles.css";

const schema = yup.object().shape({
  friends: yup
    .array()
    .of(
      yup.object().shape({
        firstName: yup.string().required("First Name is required")
      })
    )
    .min(1)
    .max(2)
});

function App() {
  const { register, handleSubmit, control, formState, clearErrors } = useForm({
    mode: "onChange",
    reValidateMode: "onChange",
    defaultValues: {
      friends: []
    },
    resolver: async (data, context, options) => {
      // you can debug your validation schema here
      const { errors, values } = await yupResolver(schema)(
        data,
        context,
        options
      );

      if (Array.isArray(errors.friends) && !formState.isSubmitted) {
        return {
          errors: {},
          values
        };
      }

      return { errors, values };
    }
  });

  console.log("formState.errors", formState.errors);

  const { fields, remove, append } = useFieldArray({
    name: "friends",
    control: control
  });

  const onSubmit = (data) => {
    console.log("submit", data);
  };

  const addFriend = () => {
    append({ firstName: "" });
    clearErrors("friends");
  };

  const removeFriend = (index) => () => {
    remove(index);
  };

  const clearFriends = () => {
    remove();
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <p>{formState?.errors?.friends?.message}</p>
      {fields.map((field, index) => {
        return (
          <fieldset key={field.id}>
            <label>
              First Name {index}:
              <input type="text" {...register(`friends.${index}.firstName`)} />
            </label>
            <button type="button" onClick={removeFriend(index)}>
              Remove
            </button>
          </fieldset>
        );
      })}

      <button type="button" onClick={addFriend}>
        Add Friend
      </button>
      <button type="button" onClick={clearFriends}>
        Clear Friends
      </button>
      <input type="submit" />
    </form>
  );
}

const root = ReactDOM.createRoot(document.getElementById("root"));
root.render(<App />);

```

**Comment by marcospassos** [comment]
Once you submit, the error never goes away:


https://user-images.githubusercontent.com/943036/216850807-0217fa80-abd3-443c-8d64-9bfb238aae67.mov

**Comment by bluebill1049** [comment]
I am just giving an example above, it's not the full implementation detail.

**Comment by marcospassos** [comment]
How can I make the error disappear without using `trigger`?

No matter what I return from the resolver, once the error is shown, the only way to make it disappear is to trigger the full field validation, which I want to avoid. I don't see any other workaround besides these I've tried with no success.

**Reply by bluebill1049** [reply]
I am covering this issue to a question as it's expected behaviour with the current library's scope.

-------

Unless the field array is valid when your schema returns a valid response, the error will be presented. React hook form doesn't memorise the errors or diff it's a root level error or individual field level error.

If a resolver return 

```
{
  errors: {
    firend: {}, // or firend: [],
  }
}
```

The `formState.errors` will persist, **until** 

```
{
  errors: {}
}
```

There is no such logic inside hook form to determine or distinguish whether each action returned it's a field array error or individual fields' error.

**Comment by Jamil-Bailony** [comment]
Hey @bluebill1049 
I faced the same issue with array field before. In addition, object field has the same issue. Error is not cleared form an `errors` object:

https://user-images.githubusercontent.com/30551053/225278068-7a68bd0e-8448-4496-96e6-ef176945f477.mp4



Check this sandbox I created mocking a real code I am working on:
https://codesandbox.io/s/react-hook-form-object-error-bug-d3eew0?file=/src/App.js

**Comment by ifrimsilviu-goosfraba** [comment]
Hello, is there any update on this issue? I'm also facing it, I've tried the following:
1. Manually trigger the re-validation of the form. It solves the problem, but it also displays all the errors of the field's inputs, which is not ideal.
2. Manually trigger the resolver when the fields of the array change. If validation passed, clear the errors of the field array. This also solves the problem, but it also hides the errors of the field's inputs, which is, again, not ideal.

**Comment by nihilenz** [comment]
Houston we have a big problem with `useFieldArray` and `errors`

**Comment by marcospassos** [comment]
Apparently, it's a bug by design 🙃

**Comment by jeffersoneagley** [comment]
Just hit a similar bug. Not really sure how I can work around it. Even calling trigger again doesn't remove the 3rd row of error. Kinda hard to use an off the shelf array item component if I have to put a state-library-specific register method on every single input within that control.

**Comment by ifrimsilviu-goosfraba** [comment]
One workaround that I've done in the recent past was to only render array errors if there aren't any items in it. 
```js
{!fields.length && renderErrors()}
```
I know it's not the ideal workaround, yet it does the job for my specific use case. Hope it helps anyone.

**Reply by vazhin** [reply]
I used this workaround, works in my case since I have an `array().min(1)` validation rule. Thanks!

**Comment by enzocussuol** [comment]
Same error here...  

I have an array of objects. I can use "remove" function from "useFieldArray()" to remove one of these objects. However, if the object i'm removing contains an error, this error is not being removed from "errors", and so i'm not able to submit the form, even if the object that contained the error no longer exist.

One solution, as discussed, is to manually trigger, as this would update "errors"... but i think this should be automatic.

**Reply by enzocussuol** [reply]
After spending some time on this, i've noticed that even triggering the entire form did not solve the issue. Error still there after removing the object and triggering form.

In fact, the problem was i was not unregistering (with the [unregister()](https://react-hook-form.com/docs/useform/unregister) method) the object before calling the remove() method. Calling unregister() and then remove() solved for me, hope it helps if someone runs into this problem.

**Comment by ChamaraWP** [comment]
Hello, I have the same issue. Even my resolver doesn't return any errors. It still shows the last state of the error I received for the `useFieldArray`. 
Here is part of the zod schema I'm facing with this issue. 
```
......
student: z
    .array(
      z.object({
        name: z.string().min(1, 'Please enter a member name'),
        subject: z.string().min(1, 'Please enter a member role'),
      })
    )
    .max(10)
    .optional(),
.....    
```

Is there any proper solution for this issue?

**Comment by sKopheK** [comment]
can confirm even in the latest version
sometimes, i have more error items for fieldarray than there is actual number of fields which might cause the issues. once i have no fields, error list is reset to 0, but once i add just a single one, all old errors appear in the error list
and even if i do edit errors manually after each add/remove, or just call clearErrors for whole fieldArray, submitting form will somehow smuggle the errors back and form won't be submitted - thus removing error for each field that's being added and filtering out invalid errors in SubmitHandler's  `onInvalid` callback parameter seems to be the solution

**Comment by salahkai** [comment]
Can confirm in the latest version when using array of objects, errors still persist, only removed when re-clicking on the submit again.

**Comment by IdanGoor** [comment]
The issue still exists 😢 I'm using version 7.55.0-next.5

**Comment by victormicco** [comment]
Goin throught the same problem nowadays
